### PR TITLE
adding commands to interact with git remotes and extract branches and tags

### PR DESCRIPTION
**Pull request by n3wscott**
**What this PR does, why we need it**:
Bringing in [buoy](https://github.com/n3wscott/buoy) to knative/test-infra.

This adds git remote fetch and extract default branch, branches, and tags.

### CODE REVIEW

**Comment by n3wscott** [comment]
/cc @vagababov 

please look at this one too! <3

**Review by vagababov — COMMENTED** [review]
/lint

:-)

File: pkg/git/git.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+const (
+	DefaultBranchRef RefType = iota
+	ReleaseBranchRef
+	ReleaseRef
+	NoRef
+)
+
+func (rt RefType) String() string {
+	if rt >= DefaultBranchRef && rt <= NoRef {
+		return [...]string{"Default Branch", "Release Branch", "Release", "No Ref"}[rt]
```
Comment (vagababov) [inline pkg/git/git.go:74]: This probably creates the array every time. Why not just global var it?

File: pkg/git/git.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	}
+	return ""
+}
+
+// BestRefFor Returns module@ref, isRelease
+func (r *Repo) BestRefFor(this semver.Version, ruleset RulesetType) (string, RefType) {
+
```
Comment (vagababov) [inline pkg/git/git.go:81]:
```suggestion
```

File: pkg/git/git.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		// Look for a Return default branch.
+		return fmt.Sprintf("%s@%s", r.Ref, r.DefaultBranch), DefaultBranchRef
+	}
+
+	// No ref found with the provided rule
+	return r.Ref, NoRef
+}
+
+func normalizeTagVersion(v string) (string, bool) {
+	if strings.HasPrefix(v, "v") {
```
Comment (vagababov) [inline pkg/git/git.go:136]: `if v[0] = 'v'` would probably work?

File: pkg/git/git_test.go [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+				t.Errorf("repo.BestRefFor() got ref = %v, want %v", got, tt.want)
+			}
+			if release != tt.release {
+				t.Errorf("repo.BestRefFor() got isRelease = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_normalizeTagVersion(t *testing.T) {
```
Comment (vagababov) [inline pkg/git/git_test.go:199]:
```suggestion
func TestNormalizeTagVersion(t *testing.T) {
```

File: pkg/git/git_test.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+				return
+			}
+			if got != tt.want {
+				t.Errorf("normalizeBranchVersion() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_tagVersion(t *testing.T) {
```
Comment (vagababov) [inline pkg/git/git_test.go:235]:
```suggestion
func TestTagVersion(t *testing.T) {
```

File: pkg/git/git_test.go [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+		t.Run(name, func(t *testing.T) {
+			got := tagVersion(tt.version)
+			if got != tt.want {
+				t.Errorf("tagVersion() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_normalizeBranchVersion(t *testing.T) {
```
Comment (vagababov) [inline pkg/git/git_test.go:267]:
```suggestion
func TestNormalizeBranchVersion(t *testing.T) {
```

File: pkg/git/git_test.go [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+				return
+			}
+			if got != tt.want {
+				t.Errorf("normalizeBranchVersion() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_branchVersion(t *testing.T) {
```
Comment (vagababov) [inline pkg/git/git_test.go:303]:
```suggestion
func TestBranchVersion(t *testing.T) {
```

File: pkg/git/ruleset.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+const (
+	AnyRule RulesetType = iota
+	ReleaseOrReleaseBranchRule
+	ReleaseRule
+	ReleaseBranchRule
+	InvalidRule
+)
+
+func (rt RulesetType) String() string {
+	return [...]string{"Any", "ReleaseOrBranch", "Release", "Branch", "Invalid"}[rt]
```
Comment (vagababov) [inline pkg/git/ruleset.go:32]: ditto here.

File: pkg/git/ruleset.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func Ruleset(rule string) RulesetType {
+	switch strings.ToLower(rule) {
```
Comment (vagababov) [inline pkg/git/ruleset.go:36]: `strings.EqualFold` is what you're looking for, I think
And also map[string]string. {lowercase to titlecase}

File: pkg/git/git.go (hunk 3 above)
Comment (n3wscott) [reply pkg/git/git.go:136]: you have to also check `if len(v) > 0 && v[0] == 'v' {` and it is easier to read as `HasPrefix`

Comment (vagababov) [reply pkg/git/git.go:136]: Do you actually expect non-empty strings at that point? It's a private function.

Comment (n3wscott) [reply pkg/git/git.go:147]: it is processing things that the git call returned, the server could return an empty tag?

Comment (n3wscott) [reply pkg/git/git.go:146]: I am gonna stick with readability over minor speed improvements, the QPS is low here.

**Review by vagababov — COMMENTED** [review]
/lgtm

**Comment by vagababov** [comment]
/hold
if you wan to fix the nits

File: pkg/git/git.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	switch ruleset {
+	case AnyRule, ReleaseOrReleaseBranchRule, ReleaseBranchRule:
+		var largest *semver.Version
+		// Look for a release branch.
+		for _, b := range r.Branches {
+			if bv, ok := normalizeBranchVersion(b); ok {
+				v, _ := semver.Make(bv)
+
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/git/git.go:122]: they fail normalizeTagVersion, they are not valid go refs from what I can tell.

File: pkg/git/git.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+			if sv, ok := normalizeTagVersion(t); ok {
+				v, _ := semver.Make(sv)
+				// Go does not understand how to fetch semver tags with pre or build tags, skip those.
+				if v.Pre != nil || v.Build != nil {
+					continue
+				}
+				if v.Major == this.Major && v.Minor == this.Minor {
+					if largest == nil || largest.LT(v) {
+						largest = &v
+					}
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/git/git.go:106]: ack, followup.

File: pkg/git/git.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+		return nil, err
+	}
+
+	for _, ref := range refs {
+		if ref.Name().IsTag() {
+			repo.Tags = append(repo.Tags, ref.Name().Short())
+		} else if ref.Name().IsBranch() {
+			repo.Branches = append(repo.Branches, ref.Name().Short())
+		} else if ref.Name() == "HEAD" { // Default branch.
+			repo.DefaultBranch = ref.Target().Short()
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/git/git.go:59]: ack, follow up

**Review by n3wscott — COMMENTED** [review]
/unhold
Thanks so much for the review @vagababov !!!!

**Pull request merged**